packages/web/components/bridge/immersive/immersive-bridge.tsx

```diff
@@ -0,0 +1,272 @@
+import { ModalCloseButton } from "~/modals";
+import { FiatOnrampSelectionModal } from "~/modals/fiat-on-ramp-selection";
+import { FiatRampsModal } from "~/modals/fiat-ramps";
+import { api } from "~/utils/trpc";
+
+enum ImmersiveBridgeScreens {
+  Asset = "0",
+  Amount = "1",
+  Review = "2",
+}
+
+const MemoizedChildren = memo(({ children }: PropsWithChildren<{}>) => {
+  return <>{children}</>;
+});
+
+export const ImmersiveBridgeFlow = ({
+  Provider,
+  children,
+}: PropsWithChildren<BridgeFlowProvider>) => {
+  const [isVisible, setIsVisible] = useState(false);
+  const [step, setStep] = useState<ImmersiveBridgeScreens>(
+    ImmersiveBridgeScreens.Asset
+  );
+  const [type, setType] = useState<"deposit" | "withdraw">("deposit");
+  const { logEvent } = useAmplitudeAnalytics();
+
+  const apiUtils = api.useUtils();
+
+  const [assetInOsmosis, setAssetInOsmosis] = useState<MinimalAsset>();
+
+  const [fiatRampParams, setFiatRampParams] = useState<{
+    fiatRampKey: FiatRampKey;
+    assetKey: string;
+  } | null>(null);
+
+  const {
+    isOpen: isFiatOnrampSelectionOpen,
+    onOpen: onOpenFiatOnrampSelection,
+    onClose: onCloseFiatOnrampSelection,
+  } = useDisclosure();
+
+  // const { isConnected, address } = useEvmWalletAccount();
+  // const { onOpenWalletSelect } = useWalletSelect();
+  // const { disconnect } = useDisconnectEvmWallet();
+
+  useLockBodyScroll(isVisible);
+
+  const onClose = () => {
+    setIsVisible(false);
+  };
+
+  const onOpen = (direction: "deposit" | "withdraw") => {
+    setIsVisible(true);
+    setType(direction);
+  };
+
+  return (
+    <Provider
+      value={{
+        startBridge: ({ direction }: { direction: "deposit" | "withdraw" }) => {
+          onOpen(direction);
+        },
+        bridgeAsset: async ({
+          anyDenom,
+          direction,
+        }: {
+          anyDenom: string;
+          direction: "deposit" | "withdraw";
+        }) => {
+          onOpen(direction);
+          setStep(ImmersiveBridgeScreens.Amount);
+
+          const fetchAssetWithRetry = async (retries = 3) => {
+            for (let attempt = 1; attempt <= retries; attempt++) {
+              try {
+                const asset = await apiUtils.edge.assets.getUserAsset.fetch({
+                  findMinDenomOrSymbol: anyDenom,
+                });
+
+                if (!asset) {
+                  console.error("Asset not found", anyDenom);
+                  return undefined;
+                }
+
+                return asset;
+              } catch (error) {
+                console.error(`Attempt ${attempt} failed:`, error);
+                if (attempt === retries) {
+                  console.error("All attempts to fetch asset failed");
+                  return undefined;
+                }
+              }
+            }
+          };
+
+          const asset = await fetchAssetWithRetry();
+
+          if (!asset) {
+            return;
+          }
+
+          setAssetInOsmosis(asset);
```

Comment:
Is this async stuff with the retry logic and catch statements necessary? Would it be possible to set a state var with the denom, then pass that as input to a query hook? Then, as long as it returns a valid asset from the query `isVisible` would return true